src/experiments/tcpsocket/api.js

```diff
@@ -0,0 +1,181 @@
+            }
+            if (receivedData && receivedData.length >= length) {
+                let promise = Promise.resolve(receivedData);
+                receivedData = null;
+                return promise;
+            }
+            console.log("(" + socketType + " waiting for " + length + " bytes)\n");
+            return new Promise(function(resolve, reject) {
+                pendingResolve = resolve;
+                wantDataLength = length;
+            });
+        },
+        waitForAnyDataAndClose() {
+            if (pendingResolve) {
+                throw new Error("only one wait allowed at a time.");
+            }
+
+            return new Promise(function(resolve, reject) {
+                pendingResolve = resolve;
+                // we may receive no data before getting close, in which case we want to
+                // return an empty array
+                receivedData = new Uint8Array();
+                wantDataAndClose = true;
+            });
+        },
+    };
+}
+
+var tcpsocket = class tcpsocket extends ExtensionAPI {
+    getAPI(context) {
+        return {
+            experiments: {
+                tcpsocket: {
+                    /** 
+                     * Send a DNS query stored in buf over a TCP socket to a 
+                     * nameserver addressed by addr
+                     */
+                    async sendDNSQuery(addr, buf) {
+                        let tcp_socket;
+                        let tcp_event_queue;
+                        let answer = new Uint8Array();
+                        try {
+                            tcp_socket = new TCPSocket(addr, 53, { binaryType: "arraybuffer" });
+                            tcp_event_queue = listenForEventsOnSocket(tcp_socket, "client");
+
+                            // Wait until the socket has opened a connection to the DNS server
+                            let nextEvent = (await tcp_event_queue.waitForEvent()).type;
+                            if (nextEvent == "open" && tcp_socket.readyState == "open") {
+                                console.log("client opened socket and readyState is open");
+                            } else {
+                                return {"error_code": 1, "data": answer};
+                            }
+                        } catch(e) {
+                            return {"error_code": 1, "data": answer};
+                        }
+
+                        try {
+                            tcp_socket.send(buf.buffer, buf.byteOffset, buf.byteLength);
+                            answer = await tcp_event_queue.waitForDataWithAtLeastLength(buf.byteLength);
+                        } catch(e) {
+                            if (e.message != "only one wait allowed at a time.") {
+                                return {"error_code": 1, "data": answer};
+                            }
+                        }
+                        return {"error_code": 0, "data": answer};
```

Comment:
(code style)
Instead of using "exit codes" disguished as error codes, either return `Promise.reject({ message: "your custom message here if you care" });` or `throw new ExtensionError("your custom message here if you care")` (from ExtensionUtils.jsm).

Then you can use `try-catch` to handle errors, and don't need to perform additional checks like `response.error_code != 0`.

This advice also applies to the `sendDNSQuery` method - instead of returning the number of written bytes, throw an error if the number of bytes does not match the expectation. After all, the extension code itself shouldn't be concerned with low-level details, all it cares about is that it can get the DNS response for a request, if any.

Reply:
Addressed by 7b2b7b97c6063f52b1cf68d925865e8a6205e44e.